File: python/nlpnewt/serialization/processor.py
# Copyright 2019 Regents of the University of Minnesota.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from argparse import ArgumentParser
from pathlib import Path
from typing import Dict, Any

from nlpnewt import processor
from nlpnewt.events import Event
from nlpnewt.processing import EventProcessor, run_processor, processor_parser
from nlpnewt.serialization.serializers import Serializer, get_serializer


@processor('nlpnewt-serializer')
class SerializationProcessor(EventProcessor):
    def __init__(self, serializer: Serializer, output_dir: str):
        self.serializer = serializer
        self.output_dir = output_dir

    def process(self, event: Event, params: Dict[str, Any]):
        name = params.get('filename', event.event_id + self.serializer.extension)
        path = Path(self.output_dir, name)
        self.serializer.event_to_file(event, path)


def main(args=None):
    parser = ArgumentParser(parents=[processor_parser()])
    parser.add_argument('serializer', help="The name of the serializer to use.")
    parser.add_argument('--output-dir', '-o', default=".",
                        help="Directory to write serialized files to.")
    ns = parser.parse_args(args)
    Serializer = get_serializer(ns.serializer)
    processor = SerializationProcessor(Serializer(), ns.output_dir)
    run_processor(processor, ns)


if __name__ == '__main__':
    main()

File: java/src/main/java/edu/umn/nlpie/mtap/processing/Processor.java
/*
 * Copyright 2019 Regents of the University of Minnesota.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.umn.nlpie.mtap.processing;


import java.lang.annotation.*;

/**
 * Annotation which marks a {@link EventProcessor} or {@link DocumentProcessor} as
 * available for service discovery registration.
 */
@Documented
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Processor {
  /**
   * The processor name. This will be the default name that the processor identifies under with
   * service discovery. It should be a valid dns path component, i.e. only consisting of
   * alphanumeric characters and dashes.
   *
   * @return String processor name.
   */
  String value();

  /**
   * A human-readable name for the processor.
   *
   * @return String human readable name like "Part of Speech Tagger"
   */
  String humanName() default "";

  /**
   * A short description of what the processor does.
   *
   * @return String description.
   */
  String description() default "";

  /**
   * The processor's entry point / main class.
   *
   * @return A string of the fully qualified class name.
   */
  String entryPoint() default "";

  /**
   * The processor's language. Defaults to "java".
   *
   * @return String identifier for the language.
   */
  String language() default "java";

  /**
   * Descriptions of the processor's parameters.
   *
   * @return Array of parameter descriptions.
   */
  ParameterDescription[] parameters() default {};

  /**
   * Descriptions of the input label indices.
   *
   * @return Array of descriptions for the label indices the processor uses.
   */
  LabelIndexDescription[] inputs() default {};

  /**
   * Descriptions of the output label indices.
   *
   * @return Array of descriptions for label indices the processor creates.
   */
  LabelIndexDescription[] outputs() default {};
}
